src/lib/Inference.hs

```diff
@@ -2629,6 +2654,53 @@ synthTopBlock block = do
   (liftExcept =<<) $ liftDictSynthTraverserM $ traverseGenericE block
 {-# SCC synthTopBlock #-}
 
+-- Given a simplified dict (an Atom of type `DictTy _` in the
+-- post-simplification IR), and a requested, more general, dict type, generalize
+-- the dict to match the more general type. This is only possible because we
+-- require that instances are polymorphic in data-role parameters.
+generalizeDict :: (EnvReader m) => Type n -> Dict n-> m n (Dict n)
```

Comment:
Why does this live in Inference? Is it because you need unification? Perhaps we should factor unification into a separate file instead, because this is weird?

Reply:
I put it here because this is where the rest of dictionary synthesis happens. `generalizeDict` is sort of like a fast-path version of dict synthesis where you don't have do figure out the whole derivation tree from scratch. You just fill in some holes. You could argue that synthesis as a whole should live in a different file though.

Reply:
Interesting take, makes sense!